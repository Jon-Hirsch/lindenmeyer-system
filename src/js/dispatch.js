let dispatch;

export default function(action) {
  return dispatch(action);
}

export function initDispatch(newDispatch) {
  dispatch = newDispatch;
}
