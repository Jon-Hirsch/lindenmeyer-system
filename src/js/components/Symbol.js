import React from 'react';
import PropTypes from 'prop-types';
import dispatch from '../dispatch';

export default function Symbol({
  name,
  operation,
  amount,
  productionRules,
  id,
  firstSymbol
}) {
  const showAmount = ['move', 'rotate'].includes(operation);
  return (
    <div className="lindenmeyer-system-symbol">
      Name:{' '}
      <input
        type="text"
        className="symbol-name"
        maxLength={1}
        value={name}
        onChange={({ target: { value } }) => updateSetting(id, 'name', value)}
      />{' '}
      | Operation:{' '}
      <select
        value={operation}
        onChange={({ target: { value } }) =>
          updateSetting(id, 'operation', value)}
      >
        <option value="none">None</option>
        <option value="move">Move Foward</option>
        <option value="rotate">Rotate</option>
        <option value="startBranch">Start Branch</option>
        <option value="endBranch">End Branch</option>
      </select>{' '}
      {showAmount && (
        <input
          type="text"
          className="symbol-amount"
          value={amount}
          onChange={({ target: { value } }) =>
            updateSetting(id, 'amount', value)}
        />
      )}{' '}
      | Production Rules:{' '}
      <input
        type="text"
        className="symbol-rules"
        value={productionRules}
        onChange={({ target: { value } }) =>
          updateSetting(id, 'productionRules', value)}
      />
      {!firstSymbol && (
        <div
          className="lindenmeyer-system-button delete-symbol-button"
          onClick={() => removeSymbol(id)}
        >
          Remove Symbol
        </div>
      )}
    </div>
  );
}

function removeSymbol(id) {
  dispatch({
    type: 'REMOVE_SYMBOL',
    id
  });
}

function updateSetting(id, setting, value) {
  dispatch({
    type: 'UPDATE_SYMBOL_SETTING',
    id,
    setting,
    value
  });
}

Symbol.propTypes = {
  name: PropTypes.string.isRequired,
  operation: PropTypes.string.isRequired,
  amount: PropTypes.string.isRequired,
  productionRules: PropTypes.string.isRequired,
  id: PropTypes.string.isRequired,
  firstSymbol: PropTypes.bool.isRequired
};
