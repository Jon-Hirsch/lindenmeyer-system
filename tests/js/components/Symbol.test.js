import React from 'react';
import Symbol from '../../../src/js/components/Symbol';
import dispatch from '../../../src/js/dispatch';

jest.mock('../../../src/js/dispatch');

describe('Symbol', () => {
  let shallowWrapper, props;

  function render() {
    return shallow(<Symbol {...props} />);
  }

  beforeEach(() => {
    dispatch.mockClear();
    props = {
      name: 'x',
      operation: 'none',
      amount: '2',
      productionRules: '+y-x-y+',
      id: '1',
      firstSymbol: true
    };

    shallowWrapper = render();
  });

  test('should match snapshot', () => {
    expect(shallowWrapper.getElement()).toMatchSnapshot();
  });

  test('should render a remove symbol button if firstSymbol is false', () => {
    props.firstSymbol = false;
    shallowWrapper = render();
    expect(shallowWrapper.find('.delete-symbol-button').length).toEqual(1);
  });

  test('should render a symbol amount if the operation is move', () => {
    props.operation = 'move';
    shallowWrapper = render();
    expect(shallowWrapper.find('.symbol-amount').length).toEqual(1);
  });

  test('should render a symbol amount if the operation is rotate', () => {
    props.operation = 'rotate';
    shallowWrapper = render();
    expect(shallowWrapper.find('.symbol-amount').length).toEqual(1);
  });

  describe('the remove symbol button', () => {
    test('should dispatch REMOVE_SYMBOL on click', () => {
      props.firstSymbol = false;
      shallowWrapper = render();
      shallowWrapper.find('.delete-symbol-button').simulate('click');
      expect(dispatch).toHaveBeenCalledWith({
        type: 'REMOVE_SYMBOL',
        id: '1'
      });
    });
  });

  describe('name field', () => {
    test('should dispatch UPDATE_SYMBOL_SETTING on change', () => {
      shallowWrapper
        .find('.symbol-name')
        .simulate('change', { target: { value: 'newValue' } });
      expect(dispatch).toHaveBeenCalledWith({
        type: 'UPDATE_SYMBOL_SETTING',
        setting: 'name',
        value: 'newValue',
        id: '1'
      });
    });
  });

  describe('operation dropdown', () => {
    test('should dispatch UPDATE_SYMBOL_SETTING on change', () => {
      shallowWrapper
        .find('select')
        .simulate('change', { target: { value: 'newValue' } });
      expect(dispatch).toHaveBeenCalledWith({
        type: 'UPDATE_SYMBOL_SETTING',
        setting: 'operation',
        value: 'newValue',
        id: '1'
      });
    });
  });

  describe('amount input', () => {
    test('should dispatch UPDATE_SYMBOL_SETTING on change', () => {
      props.operation = 'move';
      shallowWrapper = render();
      shallowWrapper
        .find('.symbol-amount')
        .simulate('change', { target: { value: 'newValue' } });
      expect(dispatch).toHaveBeenCalledWith({
        type: 'UPDATE_SYMBOL_SETTING',
        setting: 'amount',
        value: 'newValue',
        id: '1'
      });
    });
  });

  describe('rules input', () => {
    test('should dispatch UPDATE_SYMBOL_SETTING on change', () => {
      shallowWrapper
        .find('.symbol-rules')
        .simulate('change', { target: { value: 'newValue' } });
      expect(dispatch).toHaveBeenCalledWith({
        type: 'UPDATE_SYMBOL_SETTING',
        setting: 'productionRules',
        value: 'newValue',
        id: '1'
      });
    });
  });
});
