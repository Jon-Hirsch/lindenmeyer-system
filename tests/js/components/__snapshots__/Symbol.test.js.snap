// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Symbol should match snapshot 1`] = `
<div
  className="lindenmeyer-system-symbol"
>
  Name:
   
  <input
    className="symbol-name"
    maxLength={1}
    onChange={[Function]}
    type="text"
    value="x"
  />
   
  | Operation:
   
  <select
    onChange={[Function]}
    value="none"
  >
    <option
      value="none"
    >
      None
    </option>
    <option
      value="move"
    >
      Move Foward
    </option>
    <option
      value="rotate"
    >
      Rotate
    </option>
    <option
      value="startBranch"
    >
      Start Branch
    </option>
    <option
      value="endBranch"
    >
      End Branch
    </option>
  </select>
   
   
  | Production Rules:
   
  <input
    className="symbol-rules"
    onChange={[Function]}
    type="text"
    value="+y-x-y+"
  />
</div>
`;
